### ISSUE TITLE
Build libavif with encoding support on Windows

### ISSUE THREAD

**Issue by slouken** [libsdl-org/SDL_image#415]
I'd like to add support for saving images as AVIF. Could you rebuild libavif with encoder support?

**Comment by sezero** [comment]
You have to choose an encoder backend:

- rav1e: https://github.com/xiph/rav1e, encoder-only, BSD 2-Clause License, requires rust compiler which I don't have
- aom: encoder and decoder, https://aomedia.googlesource.com/aom, , BSD 2-Clause License
- svt-av1: encoder and decoder, https://gitlab.com/AOMediaCodec/SVT-AV1, BSD-3-clause clear license

Either way, the dlls will noticeably increase in size

Apart from them, do you require libyuv support in libavif? (https://chromium.googlesource.com/libyuv/libyuv/, C++)

**Comment by slouken** [comment]
I think aom is the way to go. The HDR cases are not in the libyuv fast path, but it would be a shame for us to always use the slow path.

I’ve built it here statically linked with Visual Studio and libsharpyuv and the other bells and whistles. We can use that instead, if you want.

**Comment by sezero** [comment]
> I think aom is the way to go.

OK, aom it is. We'll need to vendor it too, and drop dav1d support.

Will this be for SDL3 only?

> The HDR cases are not in the libyuv fast path, but it would be a shame for us to always use the slow path.

Hmph.. Static libstdc++ bloat will be there. Will need to vendor libyuv too

> I’ve built it here statically linked with Visual Studio and libsharpyuv and the other bells and whistles. We can use that instead, if you want.

In the interim maybe: those dlls will have all MSVC runtime and Win10 dll dependencies, no?

I'll try to build something using mingw later these days.

**Comment by slouken** [comment]
Yes, SDL3 only. The static linking with VC means we don’t have runtime dependencies.

I’ll go ahead and drop in the VC libs for now. @madebr, do you want to help with the vendoringand non-Windows builds?

**Comment by slouken** [comment]
If I remember correctly, dav1d is a faster decoder. Do we want both back ends configured?

**Comment by sezero** [comment]
> If I remember correctly, dav1d is a faster decoder. Do we want both back ends configured?

libavif allows multiple backends, don't know how it prioritizes codecs though.

**Comment by sezero** [comment]
Here are the initial libavif dll builds with encoder support: [dlls.tar.gz](https://github.com/libsdl-org/SDL_image/files/14055999/dlls.tar.gz)

Included: dav1d 1.2.1 as decoder, aom 3.6.1 as encoder, libsharpyuv from libwebp HEAD.

aom was configured without decoder support, like this:
```
-DENABLE_DOCS=0 -DENABLE_EXAMPLES=0 \
 -DENABLE_TESTDATA=0 -DENABLE_TESTS=0 -DENABLE_TOOLS=0 \
 -DBUILD_SHARED_LIBS=OFF -DCMAKE_BUILD_TYPE=Release \
 -DCONFIG_AV1_DECODER=0 -DAOM_TARGET_CPU=x86
```
IIRC, the cpu target needed explicit speciying for x86 - replace with x86_64 for x86_64

I didn't bother including libyuv yet: if you want I can do that too.

The dlls are large in size even after stripping.

**Comment by madebr** [comment]
avif has 2 configuration options to control encoding/decoding with aom: `AVIF_CODEC_AOM_ENCODE` and `AVIF_CODEC_AOM_DECODE`. Both are enabled by default.

Inside `avif.c`, the priority is dav1d > libgav1 > aom > rav1e > svg > avm

**Comment by sezero** [comment]
> AVIF_CODEC_AOM_DECODE=OFF

Will try and get back immediately

**Comment by sezero** [comment]
> > AVIF_CODEC_AOM_DECODE=OFF
> 
> Will try and get back immediately

Oops, noticed that I already disabled that option. So, these libavif builds use libaom really only for encoding

**Comment by madebr** [comment]
Isn't that good? dav1d for decoding and aom for encoding?

**Comment by sezero** [comment]
> Isn't that good? dav1d for decoding and aom for encoding?

Yes it is, and it was the intention.

Only that I didn't expect these large dll sizes only because of aom encoder inclusion (8.9M x86 dll and 13.4M x64 dll)

**Comment by slouken** [comment]
> Off-topic: I don't seem to have write permissions to our aom fork at https://github.com/libsdl-org/aom.git

Fixed!

**Comment by slouken** [comment]
> Only that I didn't expect these large dll sizes only because of aom encoder inclusion (8.9M x86 dll and 13.4M x64 dll)

I'm not surprised, the size of the aom encoder is pretty big. Remember these are optional DLLs, so users who don't want to include them in their project can leave them out.

**Comment by sezero** [comment]
> > Only that I didn't expect these large dll sizes only because of aom encoder inclusion (8.9M x86 dll and 13.4M x64 dll)
> 
> I'm not surprised, the size of the aom encoder is pretty big. Remember these are optional DLLs, so users who don't want to include them in their project can leave them out.

OK then

**Comment by sezero** [comment]
P.S.: We are currently vendoring aom 3.6.1, however 3.7.1+ has at least one CVE fix. I suggest upgrading to aom 3.7.2 or 3.8.1

**Comment by madebr** [comment]
When finished, it would also be useful to test storing avif in the test.
https://github.com/libsdl-org/SDL_image/blob/c9c809b3d488afbcc9d33228f0c50cc04ce4611c/test/main.c#L153

**Comment by slouken** [comment]
> P.S.: We are currently vendoring aom 3.6.1, however 3.7.1+ has at least one CVE fix. I suggest upgrading to aom 3.7.2 or 3.8.1

I used 3.6.1 because that is the version that is recommended with the libavif we're using. However, it looks like the API is pretty stable, we should be able to bump to a newer aom if we want.

**Comment by sezero** [comment]
Now that avif save commit https://github.com/libsdl-org/SDL_image/commit/c1600c4764c39fc15d1e1a4ffabe21919ac237a7 is in, should I push the new dlls to msvc project I you aren't doing that yourself already?

**Comment by slouken** [comment]
Sure, go for it. However I'm going to ask @madebr's help with these:
```
-- libavif: libyuv not found; libyuv-based fast paths disabled.
-- libavif: libsharpyuv not found
```

**Comment by sezero** [comment]
Pushed https://github.com/libsdl-org/SDL_image/commit/5d95dc17826a12aa6d5ec784b3f10a1643fe7388

**Comment by madebr** [comment]
> Sure, go for it. However I'm going to ask @madebr's help with these:
> 
> ```
> -- libavif: libyuv not found; libyuv-based fast paths disabled.
> -- libavif: libsharpyuv not found
> ```

libavif is looking for [libyuv](https://chromium.googlesource.com/libyuv/libyuv).
Unless we force disable it, should we vendor it?
It's c++ with lots of intrinsics.

**Comment by slouken** [comment]
I think we can vendor it. We already have libsharpyuv in webp.

**Comment by sezero** [comment]
libavif HEAD has inlined a very crippled version of libyuv under their 3rdparty directory without any cpu optimizations: We can do something similar if necessary with cpu optimizations included? And as a matter of fact most of libyuv can actually be built as C-only, at least the parts that are of interest to us: libavif's own vendored version is C-only for example.

**Comment by madebr** [comment]
[The patch](https://github.com/AOMediaCodec/libavif/commit/808bf58d11745dcc66c59c171f64ddcd020c37fa) looks very straight forward.
For libsdl-org's fork, I would just unconditionally use the "crippled version", to avoid any dependency on a system libyuv.

**Comment by madebr** [comment]
This is done in 1d91d7e39d918b34cc540309dec18e6df578a7dd

**Comment by sezero** [comment]
I made a libyuv fork and created a branch for C-only build -- here if interested: https://github.com/sezero/libyuv/tree/sezero

**Comment by madebr** [comment]
Untested, but with [`set_source_files_properties(${sources} PROPERTIES LANGUAGE C)`](https://cmake.org/cmake/help/latest/command/set_source_files_properties.html), you can override CMake's automatic language detection.

**Comment by sezero** [comment]
> Untested, but with [`set_source_files_properties(${sources} PROPERTIES LANGUAGE C)`](https://cmake.org/cmake/help/latest/command/set_source_files_properties.html), you can override CMake's automatic language detection.

OK, learnt something new about cmake - thanks.

**Comment by sezero** [comment]
Pushed https://github.com/libsdl-org/SDL_image/commit/87c811f597f9e8b881bdf9faeab344481ae2a77e which updates libavif dlls with added libyuv support included (libyuv built from my C-only fork above, and of course statically linked.)

**Comment by sezero** [comment]
Just noticed that libyuv disables cpu optimizations if targeting x86 (not x64) and `__SSE__` isn't defined and compiler isn't clang. My mingw x86 compiler is gcc (gcc 7.5.0) and doesn't enable `__SSE__` by default. That seems to be why the x86 and x64 dlls have that crazy a difference in size.

**Comment by sezero** [comment]
> Just noticed that libyuv disables cpu optimizations if targeting x86 (not x64) and `__SSE__` isn't defined and compiler isn't clang. My mingw x86 compiler is gcc (gcc 7.5.0) and doesn't enable `__SSE__` by default. That seems to be why the x86 and x64 dlls have that crazy a difference in size.

Pushed https://github.com/sezero/libyuv/commit/11df2755aa8df6e7a81d625b84dabfca04c779cb to my libyuv fork to enable SIMD functions for x86 with gcc >= 4.9 by using target attributes. Rebuilt our libavif x86 dll against it and pushed https://github.com/libsdl-org/SDL_image/commit/d5eda0410e35af0d3fc8176666112c02b4ac4e9c
